# ECM2423 - Coursework exercise

### Question 1.1: Describe how you would frame the 8-puzzle problem as a search problem.

For a problem to be framed as a search problem, more specifically a state space search, we need states, operators, a goal test and a path cost.

In the 8-puzzle problem:
- States are the locations of the tiles.
- Operators are the tile movements (left, right, up, down).
- Goal test is the given state that the puzzle aims to get to.
- Path cost is the cost per move (probably 1)

### Question 1.2: Solve the 8-puzzle problem using A*

#### 1. Briefly outline the A* algorithm.

The A* algorithm works by working out a lowest-cost path tree from the start node to the target node. For each node, A* uses a heuristic function that gives an estimate of the total cost of a path if the tree uses that node.

#### 2. Describe two admissible heuristic functions for the 8-puzzle problem and explain why they are admissible. Make sure you also explain why you chose these two heuristic functions in particular amongst all the possible ones.

You could use:
- The number of misplaced tiles: This could be used as a heuristic to ensure that this number is as low as possible so that the current search tree is optimal. This is also a very easy metric to measure and would not impact performance.
- The Manhattan distance from tile's current positions and where they need to be, same as above, it would be optimal for the algorithm to ensure these values are as small as possible. This is a fairly accurate metric to measure and gives the algorithm a clear idea of how optimal a move might be.

#### 3. Implement two versions of the A* algorithm in Python to solve the 8-puzzle using the two heuristic functions you have chosen.

In [ ]:
from aStar8Puzzle import q1_2_3

# My 8-puzzle A* code can be found in the aStar8Puzzle file, my code specific to this question can be found in the q1_2_3 method
q1_2_3()

1 2 7 
6 _ 4 
8 3 5 
  ↓
1 2 7 
6 3 4 
8 _ 5 
  ↓
1 2 7 
6 3 4 
_ 8 5 
  ↓
1 2 7 
_ 3 4 
6 8 5 
  ↓
1 2 7 
3 _ 4 
6 8 5 
  ↓
1 2 7 
3 4 _ 
6 8 5 
  ↓
1 2 _ 
3 4 7 
6 8 5 
  ↓
1 _ 2 
3 4 7 
6 8 5 
  ↓
1 4 2 
3 _ 7 
6 8 5 
  ↓
1 4 2 
3 7 _ 
6 8 5 
  ↓
1 4 2 
3 7 5 
6 8 _ 
  ↓
1 4 2 
3 7 5 
6 _ 8 
  ↓
1 4 2 
3 _ 5 
6 7 8 
  ↓
1 _ 2 
3 4 5 
6 7 8 
  ↓
_ 1 2 
3 4 5 
6 7 8 

Puzzle solved in 649 iterations
Final f cost: 14
Using Manhattan as heuristic


#### 4. Briefly discuss and compare the results given by A⋆ when using the two different heuristic functions in question 1.2.

Manhattan was a much better heuristic in my code. I used a test list of [1, 2, 7, 6, 0, 4, 8, 3, 5] which was solved in 649 iterations with manhattan and 4109 iterations with the missing tiles heuristic. Both ended up using the same steps with the same final f cost produced (14)

### Question 1.3: General solution of the 8-puzzle using A⋆

#### Write a general version of the A* algorithm (using either of the two heuristic functions described above) to solve a generic version of the 8-puzzle where the user can input any start and goal state. (Hint: can this be done for any generic pair of configurations...?)

In [10]:
from aStar8Puzzle import q1_3

# My 8-puzzle A* code can be found in the aStar8Puzzle file, my code specific to this question can be found in the q1_3 method. This code assumes that the user enters a puzzle combination that can actually be solved
q1_3()

4 1 3 
6 2 5 
7 _ 8 
  ↓
4 1 3 
6 2 5 
_ 7 8 
  ↓
4 1 3 
_ 2 5 
6 7 8 
  ↓
_ 1 3 
4 2 5 
6 7 8 
  ↓
1 _ 3 
4 2 5 
6 7 8 
  ↓
1 2 3 
4 _ 5 
6 7 8 

Puzzle solved in 75 iterations
Final f cost: 10
Using misplaced tiles as heuristic



## Question 2.1: Design and implement the Sudoko problem using Evolutionary algorithm.

#### 1. In first problem design your evolutionary algorithm addressing the following points in your design process. You should first briefly outline the how you representing Sudoko problem.

##### (a) Choose an appropriate solution space and solution representation.
##### (b) Define an appropriate fitness function.
##### (c) Define a crossover operator for the chosen representation.
##### (d) Define a mutation operator for the chosen representation.
##### (e) Decide how to initialise the population.
##### (f) Decide selection and replacement methods.
##### (g) Choose an appropriate termination criterion. 

#### 2. you should implement the evolutionary algorithm in Python to solve the Sudoku problem. 

You will need to run experiments for the three Sudoku grids provided on the ELE page, for population sizes 10, 100, 1000, 10000. Each experiment (i.e., a specific combination of grid and population size) needs to be ran 5 times (each one with a different random seed) and average performance across runs considered. In total these amount to 3 x 4 x 5 = 60 runs.

# Question 2.2: Analysis the Sudoko problem using Evolutionary algorithm.

#### This question is help guide you to analysis your results based on the following questions.
##### 1. What population size was the best?
##### 2. What do you think is the reason for your findings in question 8.a?
##### 3. Which grid was the easiest and which the hardest to solve?
##### 4. What do you think might be the reason for your findings in question 8.c?
##### 5. What further experiments do you think it may be useful to do and why?